### PR TITLE
`String.iterator()` and `String.keyValueIterator()`

### PR DESCRIPTION
**Pull request by RealyUniqueName**
- [x] eval
- [x] php
- [x] js
- [x] neko
- [x] python
- [x] lua
- [x] cpp
- [x] hl
- [x] c#
- [x] java
- [x] flash
- [x] as3

### CODE REVIEW

File: std/js/_std/Reflect.hx [hunk 1]
```diff
@@ -19,16 +19,30 @@
...
 
 	@:pure
 	public inline static function hasField( o : Dynamic, field : String ) : Bool {
 		return js.Object.prototype.hasOwnProperty.call(o, field);
 	}
 
-	@:pure
 	public static function field( o : Dynamic, field : String ) : Dynamic {
-		try return o[cast field] catch( e : Dynamic ) return null;
+		var result = try o[cast field] catch( e : Dynamic ) null;
```
Comment (nadako) [inline std/js/_std/Reflect.hx]: Not a fan of this change. We don't support this for `Array` too, so why we treat strings differently here?

Comment (RealyUniqueName) [reply std/js/_std/Reflect.hx]: > Should `String.iterator` be available through `Reflect.field(s, "iterator")` ?

> Probably yeah. Array.iterator is

@simn

Comment (Simn) [reply std/js/_std/Reflect.hx]: I thought it was... why does it work through Dynamic not but through `Reflect.field`?

Comment (nadako) [reply std/js/_std/Reflect.hx]: I think it only works if you unify with `Iterable` or something like that.

**Comment by RealyUniqueName** [comment]
https://github.com/HaxeFoundation/haxe/pull/7519/commits/6e7c2204e5604ac0164237419d4c0fa14c58a1e9 and https://github.com/HaxeFoundation/haxe/pull/7519/commits/1a92f2498e17bce864fc6c83ef23cf5ad41cff6e also fix structure access for strings on lua. E.g. `("hello":{function charAt(i:Int):Int;}).charAt(0);`

**Comment by RealyUniqueName** [comment]
I have zero experience with c/c++. 
@hughsando  @ncannasse  could you help with cpp and hl?

**Comment by Simn** [comment]
I kind of forgot about this one. We won't be able to get this ready for the RC but we should get it done for 4.0 final.

**Pull request merged**